src/hooks/UrqlUseSubscriptionWithHandler.re

```diff
@@ -0,0 +1,72 @@
+[@bs.deriving abstract]
+type useSubscriptionWithHandlerArgs = {
+  query: string,
+  [@bs.optional]
+  variables: Js.Json.t,
+};
+
+[@bs.deriving abstract]
+type useSubscriptionWithHandlerResponseJs('acc) = {
+  fetching: bool,
+  [@bs.optional]
+  data: 'acc,
+  [@bs.optional]
+  error: UrqlCombinedError.t,
+};
+
+type useSubscriptionWithHandlerResponse('acc) = {
+  fetching: bool,
+  data: option('acc),
+  error: option(UrqlCombinedError.t),
+  response: UrqlTypes.response('acc),
+};
+
+[@bs.module "urql"]
+external useSubscriptionWithHandlerJs:
+  (
+    useSubscriptionWithHandlerArgs,
+    UrqlTypes.parsedHandler('acc, 'response)
+  ) =>
+  array(useSubscriptionWithHandlerResponseJs('acc)) =
+  "useSubscription";
+
+let useSubscriptionResponseToRecordWithHandler =
+    (result: useSubscriptionWithHandlerResponseJs('acc)) => {
+  let data = result->dataGet;
+  let error = result->errorGet;
+  let fetching = result->fetchingGet;
+
+  let response =
+    switch (fetching, data, error) {
+    | (true, None, _) => UrqlTypes.Fetching
+    | (true, Some(d), _) => Data(d)
+    | (false, Some(d), _) => Data(d)
+    | (false, _, Some(error)) => Error(error)
+    | (false, None, None) => NotFound
+    };
+
+  {fetching, data, error, response};
+};
+
+let useSubscriptionWithHandler =
+    (
+      request: UrqlTypes.request('response),
+      handler: UrqlTypes.parsedHandler('acc, 'response),
+    ) => {
+  let parse = request##parse;
+  let parsedHandler =
```

Comment:
So the idea around `parsedHandler` is sort of interesting. Basically, the data that `urql` returns to you varies based on whether or not you provide a `handler` to `useSubscription`. If you **don't** provide a `handler`, `urql` will provide the raw response of executing the subscription (a `Js.Json.t` that we `parse` into `'response`). If you _do_ provide a `handler` `urql` will return whatever data structure you are using to accumulate your subscription data (referred to here as `'acc`). `parsedHandler`, realizing this, parses each `subscription` as it comes in, and then passes it on to the `handler` supplied by the user. This ensures 1) proper type inference of subscription responses as they come in, and 2) proper type inference for the `data` ('`acc') returned by `useSubscription`. You can see an example of this in the Subscription example.

Reply:
This is a cool concept, seems like some sort of reducer for subscriptions? Out of curiosity (maybe this is out of scope), how would someone have a subscription that needs to be initialized by a query?
Like if I have an `onlineUsers` query and some sort of `onlineUserUpdate` subscription that adds or removes users from that list.

Reply:
Yep, pretty much! `urql` allows you to pass a `handler` and will let you accumulate previous subscriptions with whatever data structure you want. The crux is that, when you pass a `handler`, `urql` returns the accumulator you use to maintain data, whereas when you don't, it just returns the raw subscription result, hence the need for two separate hooks here (to get the type inference) right.

In terms of a `subscription` initialized by a `query`, the only way I can think to do it is by having the component that calls `useQuery` higher up in the tree than the one that calls `useSubscription`, and passing the result of `useQuery` into the child that can then pass it into `useSubscription`. @kitten may have other thoughts, I think there's some discussion on this over in the `urql` issues.

Reply:
We were thinking of the state being combined. The initial state of the subscription being the queries initial result.

We've purposefully haven't integrated the two because that can also be done with:

- a normalised cache
- a custom hook wrapping useQuery and use Subscription
- or; local component state

So we wanted to leave this choice up to the user, since it'd otherwise feel like we were over stepping the boundary of what should and can be generalised